### PR TITLE
feat(core): add ability to split target runs across nodes

### PR DESCRIPTION
**Pull request by yharaskrik**
Adds the ability to pass in number of nodes and current node to affected run commands so that target
runs can be split evenly across nodes

## Current Behavior

Can run affected commands in parallel but only on a single machine

## Expected Behavior

Can run affected commands across more than one machine (node) so that affected commands are parallelized in more than one way

## Related Issue(s)

Fixes #

### CODE REVIEW

**Comment by minijus** [comment]
Nice! This is so needed.

Right now we use groovy scripts (inspired by [nrwl/nx-jenkins-build](https://github.com/nrwl/nx-jenkins-build#distributed-setup)) to distribute workload across multiple nodes.

Could you share the final idea how chunking should look like? Will it only return chunked lists of affected projects? Or there is something more to it?

Btw, it would be great to have possibility to run retrieval of affected projects (and chunking) without having to install node_modules locally. This would allow to run dependency installation and chunking into nodes (preparation for later CI stages) in parallel. There is an issue #4710 with more details.

**Comment by yharaskrik** [comment]
@minijus The idea with this was to not require any extra commands or custom scripts or anything. If you were running

`yarn affected:test` before you could now run `yarn affected:test --nodes=$NUM_NODES --currentNode=$CURRENT_NODE` I know CircleCI has environment variables for these im sure other systems do as well.

It is completely backwards compatible and if you do not provide the nodes it assumes it is just one machine and runs all affected. (It should also work for any target)

As for the second part of your comment, that is out of scope for this PR and something to be brought up with the Nx team. Theoretically if you were able to run Nx commands without node_modules installed then this PR would also work and you could do `yarn affected:apps --nodes=$NODES --currentNode-$CURRENT_NODE --plain`.

**Comment by FrozenPandaz** [comment]
Thank you for taking a look at this!

We're actually taking a look at this same feature. Stay tuned! :)

@yharaskrik is it okay if we close this PR? I don't think this will be the final implementation.

You can take a look at #5374 to take a peek at what it will look like :)

**Comment by yharaskrik** [comment]
Ya go ahead and close it. Wasn't sure if you guys were working on it or not and was hoping to get this in sooner rather than later.

**Comment by yharaskrik** [comment]
@FrozenPandaz one question for you, is it going to require we are using NxCloud? Or will I be able to pass in nodes/indexes from CircleCI env vars. If it is the former I will probably end up writing my own executor to chunk targets.

**Comment by vsavkin** [comment]
Hi @yharaskrik,

When it comes to distribution there are a few things to consider.

1. When distributing independent tasks, you can partition them into "bins" and run each bin on a separate node. We have a bunch of examples showing how to do it (for instance, here https://nx.dev/ci/monorepo-ci-azure). The reason why we recommend using print-affected is that it allows you to avoid the scenario where all the heavy tasks are in the same bin. If you are testing 100 projects on 3 nodes, the first 10 project can take 80% of the time. You avoid this by either randomly sorting the list of projects, so the worst case scenario is unlikely, or you weight the tasks by say running them locally and storing the timings. You can then bin them using the recorded weights. So the first node can test say 5 projects, the second 45 and the last one 50. Both are easy to implement using print-affected. This allows you to distribute independent tasks in a way where each node takes about the same amount of time.

2. When distributing tasks that aren't independent (buildable libs etc), you cannot easily use print-affected. Tasks have to run in the right order, some things can happen in parallel, but others can't. You also need to move files around from node to node, so when a particular task runs it has all the files it needs. Nx Cloud has a capability to do that and the Nx repo itself recently switched to using it.

Basically, having a coordinating service gives you the most flexibility and the best performance, but you have to have the service (e.g., Nx Cloud). Using something like print-affected is second best (cause you weight the tasks etc) and works well for independent tasks. 

We did try something like what you are suggesting. When we tried it, one of the agents would always take a lot longer than the rest, which limited the utility of the distribution. It of course depends on the workspace and might work well for your workspace. 

I'm going to close this PR for the outlined reasons.

Thank you!

**Comment by yharaskrik** [comment]
Awesome thank you for the very thorough answer @vsavkin! I really appreciate it and I never thought about the fact that the first projects take the longest that actually lines exactly with what I am seeing in our repo so I will do the "random" sort method you suggested!
